File: language-textmate/src/main/java/org/eclipse/tm4e/languageconfiguration/internal/supports/IndentRulesSupport.java
/**
 * Copyright (c) 2024 Vegard IT GmbH and others.
 *
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Initial code from https://github.com/microsoft/vscode/
 * Initial copyright Copyright (C) Microsoft Corporation. All rights reserved.
 * Initial license: MIT
 *
 * Contributors:
 * - Microsoft Corporation: Initial code, written in TypeScript, licensed under MIT license
 * - Sebastian Thomschke - translation and adaptation to Java
 */
package org.eclipse.tm4e.languageconfiguration.internal.supports;

import org.eclipse.tm4e.languageconfiguration.internal.model.IndentationRules;

/**
 * @see <a href=
 *      "https://github.com/microsoft/vscode/blob/main/src/vs/editor/common/languages/supports/indentRules.ts">
 *      github.com/microsoft/vscode/blob/main/src/vs/editor/common/languages/supports/indentRules.ts</a>
 */
public class IndentRulesSupport {

	public static final class IndentConsts {
		public static final int INCREASE_MASK = 0b00000001;
		public static final int DECREASE_MASK = 0b00000010;
		public static final int INDENT_NEXTLINE_MASK = 0b00000100;
		public static final int UNINDENT_MASK = 0b00001000;
	}

	private final IndentationRules _indentationRules;

	public IndentRulesSupport(IndentationRules indentationRules) {
		this._indentationRules = indentationRules;
	}

	public boolean shouldIncrease(final String text) {
		return _indentationRules.increaseIndentPattern.matchesFully(text);

		// if (this._indentationRules.indentNextLinePattern && this._indentationRules.indentNextLinePattern.test(text)) {
		// 	return true;
		// }
	}

	public boolean shouldDecrease(final String text) {
		return _indentationRules.decreaseIndentPattern.matchesFully(text);
	}

	public boolean shouldIndentNextLine(final String text) {
		return this._indentationRules.indentNextLinePattern != null && this._indentationRules.indentNextLinePattern.matchesFully(text);
	}

	public boolean shouldIgnore(final String text) {
		// the text matches `unIndentedLinePattern`
		return this._indentationRules.unIndentedLinePattern != null && this._indentationRules.unIndentedLinePattern.matchesFully(text);
	}

	public int getIndentMetadata(final String text) {
		int ret = 0;
		if (this.shouldIncrease(text)) {
			ret += IndentConsts.INCREASE_MASK;
		}
		if (this.shouldDecrease(text)) {
			ret += IndentConsts.DECREASE_MASK;
		}
		if (this.shouldIndentNextLine(text)) {
			ret += IndentConsts.INDENT_NEXTLINE_MASK;
		}
		if (this.shouldIgnore(text)) {
			ret += IndentConsts.UNINDENT_MASK;
		}
		return ret;
	}
}

File: language-monarch/src/main/java/io/github/rosemoe/sora/langs/monarch/languageconfiguration/support/IndentRulesSupport.kt
/*******************************************************************************
 * Copyright (c) 2015-2017 Angelo ZERR.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 * <p>
 * SPDX-License-Identifier: EPL-2.0
 * <p>
 * Contributors:
 * Angelo Zerr <angelo.zerr@gmail.com> - initial API and implementation
 * Sebastian Thomschke (Vegard IT GmbH) - add previousLineText support
 *
 ******************************************************************************/

package io.github.rosemoe.sora.langs.monarch.languageconfiguration.support

import io.github.rosemoe.sora.langs.monarch.languageconfiguration.model.IndentationRule
import io.github.rosemoe.sora.langs.monarch.utils.matchesFully

class IndentRulesSupport(
    private val indentationRules: IndentationRule
) {

    fun shouldIncrease(text: String): Boolean {
        return indentationRules.increaseIndentPattern.matchesFully(text)
    }

    fun shouldDecrease(text: String): Boolean {
        return indentationRules.decreaseIndentPattern.matchesFully(text)
    }

    fun shouldIndentNextLine(text: String): Boolean {
        return indentationRules.indentNextLinePattern?.matchesFully(text) ?: false
    }


    fun shouldIgnore(text: String): Boolean {
        return indentationRules.unIndentedLinePattern?.matchesFully(text) ?: false
    }

    fun getIndentMetadata(text: String): Int {
        var ret = 0
        if (shouldIncrease(text)) {
            ret = ret or IndentConsts.INCREASE_MASK
        }
        if (shouldDecrease(text)) {
            ret = ret or IndentConsts.DECREASE_MASK
        }
        if (shouldIndentNextLine(text)) {
            ret = ret or IndentConsts.INDENT_NEXTLINE_MASK
        }
        if (shouldIgnore(text)) {
            ret = ret or IndentConsts.UNINDENT_MASK
        }
        return ret
    }

    object IndentConsts {
        const val INCREASE_MASK = 1
        const val DECREASE_MASK = 2
        const val INDENT_NEXTLINE_MASK = 4
        const val UNINDENT_MASK = 8

    }


}